src/traces/funnelarea/plot.js

```diff
@@ -0,0 +1,306 @@
+                        .call(svgTextUtils.convertToTspans, gd);
+
+                    // position the text relative to the slice
+                    var textBB = Drawing.bBox(sliceText.node());
+                    var transform;
+
+                    var x0, x1;
+                    var y0 = Math.min(pt.BL[1], pt.BR[1]);
+                    var y1 = Math.max(pt.TL[1], pt.TR[1]);
+
+                    x0 = Math.max(pt.TL[0], pt.BL[0]);
+                    x1 = Math.min(pt.TR[0], pt.BR[0]);
+
+                    transform = getTransformToMoveInsideBar(x0, x1, y0, y1, textBB, {
+                        isHorizontal: true,
+                        constrained: true,
+                        angle: 0,
+                        anchor: 'middle'
+                    });
+
+                    sliceText.attr('transform',
+                        'translate(' + cx + ',' + cy + ')' + transform
+                    );
+                });
+            });
+
+            // add the title
+            var titleTextGroup = d3.select(this).selectAll('g.titletext')
+                .data(trace.title.text ? [0] : []);
+
+            titleTextGroup.enter().append('g')
+                .classed('titletext', true);
+            titleTextGroup.exit().remove();
+
+            titleTextGroup.each(function() {
+                var titleText = Lib.ensureSingle(d3.select(this), 'text', '', function(s) {
+                    // prohibit tex interpretation as above
+                    s.attr('data-notex', 1);
+                });
+
+                var txt = fullLayout.meta ?
+                    Lib.templateString(trace.title.text, {meta: fullLayout.meta}) :
+                    trace.title.text;
+
+                titleText.text(txt)
+                    .attr({
+                        'class': 'titletext',
+                        transform: '',
+                        'text-anchor': 'middle',
+                    })
+                .call(Drawing.font, trace.title.font)
+                .call(svgTextUtils.convertToTspans, gd);
+
+                var transform = positionTitleOutside(cd0, fullLayout._size);
+
+                titleText.attr('transform',
+                    'translate(' + transform.x + ',' + transform.y + ')' +
+                    (transform.scale < 1 ? ('scale(' + transform.scale + ')') : '') +
+                    'translate(' + transform.tx + ',' + transform.ty + ')');
+            });
+        });
+    });
+};
+
+function move(a) {
+    return 'm' + a[0] + ',' + a[1];
+}
+
+function line(a, b) {
+    var dx = b[0] - a[0];
+    var dy = b[1] - a[1];
+
+    return 'l' + dx + ',' + dy;
+}
+
+function getBetween(a, b) {
+    return [
+        0.5 * (a[0] + b[0]),
+        0.5 * (a[1] + b[1])
+    ];
+}
+
+function setCoords(cd) {
+    if(!cd.length) return;
+
+    var cd0 = cd[0];
+
+    var aspectratio = cd0.trace.aspectratio;
+    var h = cd0.trace.baseratio;
+    if(h > 0.999) h = 0.999;
+    var h2 = Math.pow(h, 2);
```

Comment:
That's a cheeky line. What happens when `h=1`?

Reply:
... ha I see, `v0` blows up. Is there a way to handle the `h=1` case separately?

Reply:
The origin moves to `infinity` which should be avoided. By the way `h=1` or `h=0.999` could be used to generate rectangles.

Reply:
> ... ha I see, `v0` blows up. Is there a way to handle the `h=1` case separately?

Yes. That might be a better option later on the road. But I think for the moment we could/should skip that.

Reply:
Sure, skipping is fine for now! 

Make sure to add a `TODO` above this line though.

Reply:
Done in https://github.com/plotly/plotly.js/pull/3876/commits/7e6b9aee160110a8d17f3be746f6ccf44531501d.